crates/ruff_linter/src/rules/fastapi/rules/fastapi_unused_path_parameter.rs

```diff
@@ -0,0 +1,198 @@
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self {
+            arg_name,
+            function_name,
+            arg_name_already_used: arg_name_is_pos_only_arg,
+        } = self;
+        if *arg_name_is_pos_only_arg {
+            format!(
+                "Path parameter `{arg_name}` in route path, but appears as a positional-only argument in function `{function_name}`. Consider making it a regular argument."
+            )
+        } else {
+            format!(
+                "Path parameter `{arg_name}` in route path but not in function `{function_name}`."
+            )
+        }
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        let Self { arg_name, .. } = self;
+        if self.arg_name_already_used {
+            None
+        } else {
+            Some(format!("Add `{arg_name}` to function signature"))
+        }
+    }
+}
+
+/// Returns a vector of path parameters and their ranges in the route path.
+/// The string is just the name of the path parameter.
+/// The range includes the curly braces.    
+fn extract_path_params_from_route(input: &str) -> Vec<(String, Range<usize>)> {
+    // We ignore text after a colon, since those are path convertors
+    // See also: https://fastapi.tiangolo.com/tutorial/path-params/?h=path#path-convertor
+    let re = Regex::new(r"\{([^:}]+).*?\}").unwrap();
+
+    // Collect all matches and return them as a vector of strings
+    re.captures_iter(input)
+        .filter_map(|cap| {
+            if let (Some(name_match), Some(full_match)) = (cap.get(1), cap.get(0)) {
+                Some((name_match.as_str().trim().to_string(), full_match.range()))
+            } else {
+                None
+            }
+        })
+        .collect()
+}
+
+pub(crate) fn fastapi_unused_path_parameter(
+    checker: &mut Checker,
+    function_def: &ast::StmtFunctionDef,
+) {
+    if !checker.semantic().seen_module(Modules::FASTAPI) {
+        return;
+    }
+    if !is_fastapi_route(function_def, checker.semantic()) {
+        return;
+    }
+
+    // Get the route path from the decorator
+    let route_decorator = function_def
+        .decorator_list
+        .iter()
+        .find_map(|decorator| is_fastapi_route_decorator(decorator, checker.semantic()));
+
+    let Some(route_decorator) = route_decorator else {
+        return;
+    };
+
+    let path_arg = route_decorator.arguments.args.first();
+    let Some(path_arg) = path_arg else {
+        return;
+    };
+    // Lets path_arg go out of scope so we can reuse checker later
+    let diagnostic_range = path_arg.range();
+    // We can't really handle anything other than string literals
+    let path = match path_arg.as_string_literal_expr() {
+        Some(path_arg) => path_arg.value.to_string(),
+        None => return,
+    };
+
+    let path_params = extract_path_params_from_route(&path);
```

Comment:
It would be nice if we can avoid returning a `Vec` from `extract_path_params_from_route` to reduce the number of allocations (allocations are rather expensive). 

I think it could be possible but you probably have to write your own iterator type

```rust
fn extract_path_params_from_route(input: &str) -> PathParamsIter {
	PathParamsIter {
		rest: input,
		offset: TextSize::default()
	}
}

struct PathParamsIter<'a> {
	rest: &'a str,
	offset: TextSize,
}

impl<'a> Iterator for PathParamsIter<'a> {
	type Item = (&'a str, TextRange);

	fn next(&mut self) {
	 // TODO do we have to split off the `{`, `:` and `}` from name etc?
		let (before_curly, after_curly) = self.rest.split_once('{')?;
		let (name, after_colon) = after_curly.split_once(':')?;
		let (up_to_curly, rest) = after_colon.split_once('}')?;
		let offset = self.offset;
		let len = self.rest.text_len() - rest.text_len();

		self.offset += len;
		self.rest = rest;

		Some(name.trim(), TextRange::at(offset, len)
	}
}

impl FusedIterator for PathParamsIter<'_> {}
```